bincopy.py

```diff
@@ -721,6 +729,67 @@ def add_ihex(self, records, overwrite=False):
                     record,
                     type_))
 
+    def add_ti_txt(self, records, overwrite=False):
+        """Add given TI-TXT records. Set `overwrite` to ``True`` to allow
+        already added data to be overwritten.
+
+        """
+
+        offset = -1
+        last_len = TI_TXT_BYTES_PER_LINE
+        eol = 0
+        segment = None
+
+        for record in StringIO(records):
+            add_segment = False
+
+            # Check if there's data after EOL
+            if eol:
+                raise Error("bad file terminator")
+
+            # Check if the line is EOL
+            if record[0] == 'q':
+                eol = 1
+                add_segment = True
+
+            # Check if record is offset
+            elif record[0] == '@':
+                try:
+                    offset = int(record[1:], 16)
+                    last_len = TI_TXT_BYTES_PER_LINE
+                except ValueError:
+                    raise Error("bad address")
+                add_segment = True
+
+            else:
+                # Try to decode data
+                try:
+                    chunk = bytearray(binascii.unhexlify(record.strip().replace(" ", "")))
+                except (TypeError, binascii.Error):
+                    raise Error("bad data: {}".format(record))
+
+                if offset < 0:
+                    raise Error("missing offset")
+
+                if last_len != TI_TXT_BYTES_PER_LINE:
+                    raise Error("bad record length ({})".format(last_len))
+
+                last_len = len(chunk)
+
+                if segment:
+                    segment.add_data(offset, offset+last_len, chunk, False)
+                else:
+                    segment = _Segment(offset, offset+last_len, chunk)
+
+                offset += last_len
+
+            if add_segment and segment:
+                self._segments.add(segment, overwrite)
+                segment = None
+
+        if not eol:
+            raise Error("missing file terminator")
+
```

Comment:
I suggest something like this instead. It removes the need for keeping track of the length of the previous data line, making the code easier to understand.

``` python
    def add_ti_txt(self, records, overwrite=False):
        """Add given TI-TXT records. Set `overwrite` to ``True`` to allow
        already added data to be overwritten.

        """

        address = None
        eol_found = False

        for record in StringIO(records):
            # Abort if data is found after end of line.
            if eol_found:
                raise Error("bad file terminator")

            if record[0] == 'q':
                # Check if the line is EOL.
                eol_found = True
            elif record[0] == '@':
                # Check if record is an address.
                try:
                    address = int(record[1:], 16)
                except ValueError:
                    raise Error("bad section address")
            else:
                # Try to decode data
                try:
                    data = bytearray(binascii.unhexlify(record.strip().replace(" ", "")))
                except (TypeError, binascii.Error):
                    raise Error("bad data: {}".format(record))

                if address is None:
                    raise Error("missing section address")

                size = len(data)

                if size > TI_TXT_BYTES_PER_LINE:
                    raise Error("line too long")
                
                segment = _Segment(address, address + size, data)
                self._segments.add(segment, overwrite)
                
                if size == TI_TXT_BYTES_PER_LINE:
                    address += size
                else:
                    address = None

        if not eol_found:
            raise Error("missing file terminator")
```

Reply:
Your code is easier to read, that's for sure. I still need to keep the length of last line in order to get a more descriptive error message in case there are too short records in the middle of a section. Also tests are updated to check for this a bit more thoroughly.

Reply:
The error messages could be improved further. It's usually a good habit to include the condition that failed. For example, replacing `Error("bad record length")` with `Error("expected data of length 16 or shorter, but got "{}" (length {})".format(record.strip(), size))` would help the user identify the problem.

Reply:
I agree on everything you wrote. Let's leave the messages as is for now.

Back to you previous comment about keeping the length of the last line for a better error message. Finding a 16 bytes record directly after short record can mean two things; either the short record is missing data, or a section address is missing after the short record. As we can't know which, I prefer to make the code as simple as possible and simply raise the section address missing message.

Reply:
Ok. Do you think this is now ready to be merged or would you like me to still make some changes? 

I'm quite new with GitHub so (even after going through the documentation) I'm not sure if I still need to do something.

Reply:
I would prefer raising an error with "missing section address" instead of "bad record length" if a 16 bytes record is found directly after a short record, as explained in my previous comment. Other than that you are done.

Reply:
Fair enough. I changed the error message like you said. I also changed the end of `as_ti_txt` to be consistent with `as_srec` and `as_ihex`.